File: backend/src/schemas/prompt_template.py
"""
Pydantic schemas for PromptTemplate API endpoints.

These schemas define the structure for request/response validation
and serialization for all prompt template-related API operations.
"""

from datetime import datetime
from typing import Optional, Dict, Any, List
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class PromptTemplateBase(BaseModel):
    """Base schema with common prompt template fields."""

    name: str = Field(..., min_length=1, max_length=255, description="Template name")
    description: Optional[str] = Field(None, description="Template description")


class PromptTemplateCreate(PromptTemplateBase):
    """Schema for creating a new prompt template."""

    prompts: List[str] = Field(
        ...,
        min_length=1,
        description="Array of prompt strings"
    )
    is_favorite: bool = Field(False, description="Whether this template is marked as favorite")
    tags: Optional[List[str]] = Field(default_factory=list, description="Tags for organization")

    @field_validator("prompts")
    @classmethod
    def validate_prompts(cls, v: List[str]) -> List[str]:
        """Validate prompts - filter out empty strings and require at least one non-empty prompt."""
        # Filter out empty/whitespace-only prompts
        valid_prompts = [p.strip() for p in v if p.strip()]
        if not valid_prompts:
            raise ValueError("At least one non-empty prompt is required")
        return valid_prompts

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, v: Optional[List[str]]) -> List[str]:
        """Validate and clean tags."""
        if v is None:
            return []
        # Filter out empty tags and strip whitespace
        return [tag.strip() for tag in v if tag.strip()]


class PromptTemplateUpdate(BaseModel):
    """Schema for updating an existing prompt template."""

    name: Optional[str] = Field(None, min_length=1, max_length=255, description="Template name")
    description: Optional[str] = Field(None, description="Template description")
    prompts: Optional[List[str]] = Field(None, min_length=1, description="Array of prompt strings")
    is_favorite: Optional[bool] = Field(None, description="Favorite status")
    tags: Optional[List[str]] = Field(None, description="Tags for organization")

    @field_validator("prompts")
    @classmethod
    def validate_prompts(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Validate prompts if provided."""
        if v is None:
            return None
        valid_prompts = [p.strip() for p in v if p.strip()]
        if not valid_prompts:
            raise ValueError("At least one non-empty prompt is required")
        return valid_prompts

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Validate and clean tags if provided."""
        if v is None:
            return None
        return [tag.strip() for tag in v if tag.strip()]


class PromptTemplateResponse(PromptTemplateBase):
    """Schema for prompt template response."""

    id: UUID = Field(..., description="Unique template identifier (UUID)")
    prompts: List[str] = Field(..., description="Array of prompt strings")
    is_favorite: bool = Field(..., description="Whether this template is marked as favorite")
    tags: List[str] = Field(default_factory=list, description="Tags for organization")
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: datetime = Field(..., description="Record last update timestamp")

    model_config = {
        "from_attributes": True,
    }


class PromptTemplateListResponse(BaseModel):
    """Schema for paginated list of prompt templates."""

    data: list[PromptTemplateResponse] = Field(..., description="List of prompt templates")
    pagination: Dict[str, Any] = Field(..., description="Pagination metadata")


class PromptTemplateExport(BaseModel):
    """Schema for exporting prompt templates to JSON."""

    version: str = Field("1.0", description="Export format version")
    templates: List[PromptTemplateResponse] = Field(..., description="List of templates to export")
    exported_at: datetime = Field(..., description="Export timestamp")


class PromptTemplateImport(BaseModel):
    """Schema for importing prompt templates from JSON."""

    version: str = Field(..., description="Import format version")
    templates: List[PromptTemplateCreate] = Field(..., min_length=1, description="List of templates to import")

    @field_validator("version")
    @classmethod
    def validate_version(cls, v: str) -> str:
        """Validate import version compatibility."""
        if v not in ["1.0"]:
            raise ValueError(f"Unsupported import version: {v}. Supported versions: 1.0")
        return v

File: frontend/src/types/promptTemplate.ts
/**
 * Prompt Template types and interfaces.
 *
 * These types match the backend Pydantic schemas for prompt template management.
 */

/**
 * Prompt template for saving and reusing prompt series for steering experiments.
 * Matches backend PromptTemplate model.
 */
export interface PromptTemplate {
  /** UUID identifier */
  id: string;
  /** Template name */
  name: string;
  /** Optional description */
  description?: string;
  /** Array of prompt strings */
  prompts: string[];
  /** Whether this template is marked as favorite */
  is_favorite: boolean;
  /** Tags for organization */
  tags: string[];
  /** Creation timestamp */
  created_at: string;
  /** Last update timestamp */
  updated_at: string;
}

/**
 * Data required to create a new prompt template.
 * Matches backend PromptTemplateCreate schema.
 */
export interface PromptTemplateCreate {
  /** Template name (required) */
  name: string;
  /** Optional description */
  description?: string;
  /** Array of prompt strings (at least one required) */
  prompts: string[];
  /** Whether this template is marked as favorite (default: false) */
  is_favorite?: boolean;
  /** Tags for organization */
  tags?: string[];
}

/**
 * Data for updating an existing prompt template.
 * All fields are optional. Matches backend PromptTemplateUpdate schema.
 */
export interface PromptTemplateUpdate {
  /** Template name */
  name?: string;
  /** Description */
  description?: string;
  /** Array of prompt strings */
  prompts?: string[];
  /** Whether this template is marked as favorite */
  is_favorite?: boolean;
  /** Tags for organization */
  tags?: string[];
}

/**
 * Pagination metadata for list responses.
 */
export interface PaginationMetadata {
  /** Current page number (1-indexed) */
  page: number;
  /** Items per page */
  limit: number;
  /** Total number of items */
  total: number;
  /** Total number of pages */
  total_pages: number;
  /** Whether there is a next page */
  has_next: boolean;
  /** Whether there is a previous page */
  has_prev: boolean;
}

/**
 * API response for list of prompt templates.
 */
export interface PromptTemplateListResponse {
  /** Array of prompt templates */
  data: PromptTemplate[];
  /** Pagination information */
  pagination: PaginationMetadata;
}

/**
 * Export data structure for prompt templates.
 */
export interface PromptTemplateExport {
  /** Export format version */
  version: string;
  /** ISO timestamp of export */
  exported_at: string;
  /** Array of templates */
  templates: PromptTemplate[];
}

/**
 * Import result from importing prompt templates.
 */
export interface PromptTemplateImportResult {
  /** Number of templates created */
  created: number;
  /** Number of templates updated */
  updated: number;
  /** Number of templates skipped (duplicates) */
  skipped: number;
  /** Total number of templates processed */
  total: number;
}

/**
 * Query parameters for listing prompt templates.
 */
export interface PromptTemplateListParams {
  /** Page number (1-indexed) */
  page?: number;
  /** Items per page (1-100) */
  limit?: number;
  /** Search query for name or description */
  search?: string;
  /** Filter by favorite status */
  is_favorite?: boolean;
  /** Filter by tag */
  tag?: string;
  /** Sort by field (default: 'created_at') */
  sort_by?: 'name' | 'created_at' | 'updated_at';
  /** Sort order (default: 'desc') */
  order?: 'asc' | 'desc';
}
